src/vs/workbench/contrib/terminal/browser/terminalVoice.ts

```diff
@@ -0,0 +1,208 @@
+	private _ghostTextMarker: IMarker | undefined;
+	private static _instance: TerminalVoiceSession | undefined = undefined;
+	private _acceptTranscriptionScheduler: RunOnceScheduler | undefined;
+	static getInstance(instantiationService: IInstantiationService): TerminalVoiceSession {
+		if (!TerminalVoiceSession._instance) {
+			TerminalVoiceSession._instance = instantiationService.createInstance(TerminalVoiceSession);
+		}
+
+		return TerminalVoiceSession._instance;
+	}
+	private _cancellationTokenSource: CancellationTokenSource | undefined;
+	private readonly _disposables: DisposableStore;
+	constructor(
+		@ISpeechService private readonly _speechService: ISpeechService,
+		@ITerminalService private readonly _terminalService: ITerminalService,
+		@IConfigurationService private readonly _configurationService: IConfigurationService,
+	) {
+		super();
+		this._register(this._terminalService.onDidChangeActiveInstance(() => this.stop()));
+		this._register(this._terminalService.onDidDisposeInstance(() => this.stop()));
+		this._disposables = this._register(new DisposableStore());
+	}
+
+	async start(): Promise<void> {
+		this.stop();
+		let voiceTimeout = this._configurationService.getValue<number>(AccessibilityVoiceSettingId.SpeechTimeout);
+		if (!isNumber(voiceTimeout) || voiceTimeout < 0) {
+			voiceTimeout = SpeechTimeoutDefault;
+		}
+		this._acceptTranscriptionScheduler = this._disposables.add(new RunOnceScheduler(() => {
+			this._sendText();
+			this.stop();
+		}, voiceTimeout));
+		this._cancellationTokenSource = new CancellationTokenSource();
+		this._register(toDisposable(() => this._cancellationTokenSource?.dispose(true)));
+		const session = await this._speechService.createSpeechToTextSession(this._cancellationTokenSource?.token, 'terminal');
+
+		this._disposables.add(session.onDidChange((e) => {
+			if (this._cancellationTokenSource?.token.isCancellationRequested) {
+				return;
+			}
+			switch (e.status) {
+				case SpeechToTextStatus.Started:
+					if (!this._decoration) {
+						this._createDecoration();
+					}
+					break;
+				case SpeechToTextStatus.Recognizing: {
+					this._updateInput(e);
+					this._renderGhostText(e);
+					if (voiceTimeout > 0) {
+						this._acceptTranscriptionScheduler!.cancel();
+					}
+					break;
+				}
+				case SpeechToTextStatus.Recognized:
+					this._updateInput(e);
+					if (voiceTimeout > 0) {
+						this._acceptTranscriptionScheduler!.schedule();
+					}
+					break;
+				case SpeechToTextStatus.Stopped:
+					this.stop();
+					break;
+			}
+		}));
+	}
+	stop(send?: boolean): void {
+		this._setInactive();
+		if (send) {
+			this._acceptTranscriptionScheduler!.cancel();
+			this._sendText();
+		}
+		this._marker?.dispose();
+		this._ghostTextMarker?.dispose();
+		this._ghostText?.dispose();
+		this._ghostText = undefined;
+		this._decoration?.dispose();
+		this._decoration = undefined;
+		this._cancellationTokenSource?.cancel();
+		this._disposables.clear();
```

Comment:
We have `_disposables`, so this._ghostText and the like could be registered on that when it's created?